### PR TITLE
feat: Add statistical significance testing to evaluation comparisons …

### LINKED ISSUE #66: [Feature]: Statistical significance testing in evaluation comparisons (PRD FR-E06)
## Problem Statement

The PRD (FR-E06, priority "Should Have") requires: "Baseline vs. trained comparison with statistical significance." The TRD references this as a core evaluation capability. The user story US-12 maps to it with test reference `test_evaluation.py::test_compare`.

However, `Evaluator.compare()` (evaluator.py line 203) only computes raw deltas:

```python
deltas = {}
for key in baseline:
    if isinstance(baseline.get(key), (int, float)) and isinstance(trained.get(key), (int, float)):
        deltas[key] = trained[key] - baseline[key]
```

There is no:
- Confidence interval computation
- Statistical significance test (paired t-test, Wilcoxon, bootstrap)
- Effect size measurement (Cohen's d)
- Multiple comparison correction (Bonferroni)

This means a user seeing "robustness_delta: +0.018" has no way to know if that's a statistically meaningful improvement or random noise from training variance. For a platform targeting researchers (Persona: Dr. Priya) and compliance officers (Persona: Sarah), this is a credibility gap.

**Relevant files:**
- `nightmarenet/evaluation/evaluator.py` lines 203-242 (`compare()` method)
- `docs/architecture/PRD.md` FR-E06

## Proposed Solution

Add statistical testing to the comparison output:

1. Run evaluation N times (configurable, default 5) with different random seeds to get a distribution
2. Compute paired t-test or bootstrap confidence interval for each metric delta
3. Report: delta, p-value, 95% CI, effect size
4. Mark deltas as "significant" only if p < 0.05

For single-run comparisons (cheaper), at minimum compute bootstrap CI from the per-strength scores (the robustness sweep already evaluates at 9 strengths, giving 9 paired observations).

**Files affected:**
- `nightmarenet/evaluation/evaluator.py` - enhance `compare()` with significance testing
- `configs/default.yaml` - `evaluation.significance_runs: 5` (optional)
- Dependency: `scipy.stats` (already available via scikit-learn transitive dep)

## Acceptance Criteria

- [ ] `compare()` output includes p-value and confidence interval for robustness delta
- [ ] Significance determined at configurable alpha (default 0.05)
- [ ] Method documented: which test (bootstrap CI or paired t-test), what N
- [ ] Markdown report includes "statistically significant" / "not significant" verdict
- [ ] Works with single-run data (bootstrap from per-strength scores) and multi-run data
- [ ] Test: two identical models produce p > 0.05 (not significant)

## Scope / Difficulty Estimate

Medium. ~3-4 hours. scipy.stats has the primitives; the work is integrating them into the comparison flow and deciding on methodology.

### PR DESCRIPTION
**Pull request by mrunmayeekokitkar**
## Summary

Add statistical significance testing for robustness metric comparisons using bootstrap confidence intervals and paired t-tests, with configurable significance thresholds and reporting support.

## Motivation

Closes #66

Robustness comparisons previously reported only aggregate metric differences, making it difficult to determine whether observed improvements were statistically meaningful. This change adds significance testing based on per-strength robustness scores, enabling confidence intervals, p-values, and significance verdicts to be included in comparison reports.

### `nightmarenet/evaluation/evaluator.py`

- Added `_bootstrap_ci()` to compute bootstrap confidence intervals and paired t-test p-values from per-strength robustness scores
- Enhanced `compare()` to perform statistical significance testing for robustness comparisons
- Added configurable `significance_alpha` parameter (default: `0.05`)
- Updated `generate_report()` to include:
  - Mean robustness difference
  - 95% confidence interval
  - Paired t-test p-value
  - Statistical significance verdict
- Added handling for `NaN` p-values caused by zero-variance (identical) data

### `tests/test_evaluation.py`

- Added test verifying identical models produce `p > 0.05` (not statistically significant)
- Added test verifying different models produce `p < 0.05` (statistically significant)
- Added test for insufficient data handling
- Added test for custom `significance_alpha` thresholds

## Acceptance Criteria

- [x] Add bootstrap CI and paired t-test for robustness metric comparisons
- [x] Compute p-value, 95% CI, and significance flag from per-strength scores
- [x] Add configurable `significance_alpha` parameter (default `0.05`)
- [x] Update markdown report to include significance verdict
- [x] Add tests for identical models (`p > 0.05`) and different models (`p < 0.05`)
- [x] Handle `NaN` p-values from zero-variance data

## Type

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] Feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would break existing behavior)
- [ ] Refactor (no functional change)
- [ ] Documentation
- [x] Tests
- [ ] Infrastructure (CI/CD, Docker, tooling)

---

## Pre-submission Requirements

> These are mandatory. PRs missing any item will not be reviewed.

- [x] I have **starred** this repository ⭐
- [x] I have **followed** [@Adit-Jain-srm](https://github.com/Adit-Jain-srm)
- [x] I have read [CONTRIBUTING.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/CONTRIBUTING.md) in full

## Quality Checklist

- [x] `ruff check nightmarenet/ scripts/ tests/` — 0 errors
- [x] `mypy nightmarenet/ --ignore-missing-imports` — no new errors (run on Python 3.12)
- [x] `pytest tests/` — all tests pass
- [x] New functionality has corresponding tests
- [x] Commit messages follow [Conventional Commits](https://www.conventionalcommits.org/) (`feat:`, `fix:`, `docs:`, `test:`, `refactor:`, `chore:`)
- [x] No `from __future__ import annotations` added under `nightmarenet/api/`
- [x] No new imports of hosted-only libraries (`sqlalchemy`, `redis`, `celery`) in `nightmarenet/`

## Documentation

- [x] No documentation needed
- [ ] README.md updated
- [ ] Docstrings added/updated (Google style, public APIs only)
- [ ] `docs/` updated (architecture, research, or API docs)
- [ ] Config comments updated (`configs/default.yaml`)
- [ ] CHANGELOG entry added (if user-facing change)

## AI Disclosure

- [ ] This PR is entirely human-written
- [x] This PR includes AI-assisted code (Copilot, ChatGPT, Claude, etc.) — I have reviewed every line and can explain it.

## Security Considerations

- [x] Not applicable (no security-sensitive changes)
- [ ] User input is validated before use
- [x] No secrets, keys, or credentials in code or comments
- [x] CORS / auth / rate limiting behavior unchanged (or explicitly documented)
- [ ] If this is a security fix, see [SECURITY.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/SECURITY.md) — do NOT describe the vulnerability publicly until patched

## Screenshots (if UI/UX change)

| View | Screenshot |
|------|-----------|
| Desktop (dark mode) | N/A |
| Desktop (light mode) | N/A |
| Mobile (375px) | N/A |

## Breaking Changes

N/A

---

<details>
<summary>Reviewer notes (optional)</summary>

- Statistical testing uses the per-strength robustness scores already produced during evaluation.
- Bootstrap confidence intervals are computed using 10,000 bootstrap samples.
- A comparison is considered statistically significant only when the 95% confidence interval excludes zero **and** the paired t-test p-value is below the configured `significance_alpha`.
- Zero-variance data is handled gracefully by treating `NaN` p-values as non-significant.

</details>

### CODE REVIEW

**Comment by Adit-Jain-srm** [comment]
**One issue to fix:**

**Tests duplicate the implementation instead of importing it.**

`tests/test_evaluation.py` copy-pastes the entire `_bootstrap_ci` function (70 lines) instead of importing from `nightmarenet.evaluation.evaluator`. This means:
- If someone fixes a bug in the real function, the test still passes on the stale copy
- The actual `_bootstrap_ci` in `evaluator.py` is never tested

Fix:
```python
from nightmarenet.evaluation.evaluator import _bootstrap_ci
```

Remove the duplicated function body from the test file.

**Non-blocking observations:**

1. **scipy is not in pyproject.toml dependencies.** It works because scikit-learn pulls it in transitively, but if someone installs without sklearn, `from scipy import stats` will crash. Either add `scipy>=1.10.0` to `[project.dependencies]` or wrap the import with a graceful fallback.

2. **Bootstrap is not seeded.** `np.random.choice` uses global state, making test results non-deterministic. Consider accepting an optional `rng` parameter:
   ```python
   rng = np.random.default_rng(seed)
   indices = rng.choice(n, size=n, replace=True)
   ```
   This also prevents the bootstrap from interfering with training reproducibility (issue #64).

3. **Using inverse perplexity as the paired metric is unusual.** The issue proposed using per-strength accuracy scores directly. Inverse perplexity is a valid proxy but worth documenting WHY in a code comment (perplexity is lower-is-better, so 1/ppl makes it higher-is-better for the paired test to work correctly). A reviewer unfamiliar with the codebase won't know this.

4. **No config file update.** Issue mentions `configs/default.yaml` should get `evaluation.significance_runs: 5` or similar. The PR adds `significance_alpha` in code but doesn't document it in the config. Add a comment to `configs/default.yaml` under the evaluation section.

**Summary of required changes:**
- [ ] Import `_bootstrap_ci` in tests instead of duplicating it
- [ ] Add `scipy` to pyproject.toml dependencies (or document it)
- [ ] Seed the bootstrap RNG for deterministic tests
- [ ] Add `significance_alpha` comment to configs/default.yaml

**Review by Adit-Jain-srm — CHANGES_REQUESTED** [review]
Please implement these and it's good to merge @mrunmayeekokitkar

**Review by Adit-Jain-srm — APPROVED** [review]
LGTM!
